azure/services/natgateways/natgateways.go

```diff
@@ -55,119 +57,61 @@ func New(scope NatGatewayScope) *Service {
 // Reconcile gets/creates/updates a nat gateway.
 // Only when the Nat Gateway 'Name' property is defined we create the Nat Gateway: it's opt-in.
 func (s *Service) Reconcile(ctx context.Context) error {
-	ctx, _, done := tele.StartSpanWithLogger(ctx, "natgateways.Service.Reconcile")
-	defer done()
+	ctx, span := tele.Tracer().Start(ctx, "natgateways.Service.Reconcile")
+	defer span.End()
 
-	if !s.Scope.Vnet().IsManaged(s.Scope.ClusterName()) {
-		s.Scope.V(4).Info("Skipping nat gateways reconcile in custom vnet mode")
-		return nil
-	}
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureServiceReconcileTimeout)
+	defer cancel()
 
+	// We go through the list of NatGatewaySpecs to reconcile each one, independently of the result of the previous one.
+	// If multiple errors occur, we return the most pressing one
+	// order of precedence is: error creating -> creating in progress -> created (no error)
+	var result error
 	for _, natGatewaySpec := range s.Scope.NatGatewaySpecs() {
-		existingNatGateway, err := s.getExisting(ctx, natGatewaySpec)
-
-		switch {
-		case err != nil && !azure.ResourceNotFound(err):
-			return errors.Wrapf(err, "failed to get nat gateway %s in %s", natGatewaySpec.Name, s.Scope.ResourceGroup())
-		case err == nil:
-			// nat gateway already exists
-			s.Scope.V(4).Info("nat gateway already exists", "nat gateway", natGatewaySpec.Name)
-			natGatewaySpec.Subnet.NatGateway.ID = existingNatGateway.ID
-
-			if existingNatGateway.NatGatewayIP.Name == natGatewaySpec.NatGatewayIP.Name {
-				// Skip update for Nat Gateway as it exists with expected values
-				s.Scope.V(4).Info("Nat Gateway exists with expected values, skipping update", "nat gateway", natGatewaySpec.Name)
-				natGatewaySpec.Subnet.NatGateway = *existingNatGateway
-				s.Scope.SetSubnet(natGatewaySpec.Subnet)
-				continue
-			} else {
-				s.Scope.V(2).Info("updating NAT gateway IP name to match the spec", "old name", existingNatGateway.NatGatewayIP.Name, "desired name", natGatewaySpec.NatGatewayIP.Name)
-			}
-		default:
-			// nat gateway doesn't exist but its name was specified in the subnet, let's create it
-			s.Scope.V(2).Info("nat gateway doesn't exist yet, creating it", "nat gateway", natGatewaySpec.Name)
-		}
-
-		natGatewayToCreate := network.NatGateway{
-			Location: to.StringPtr(s.Scope.Location()),
-			Sku:      &network.NatGatewaySku{Name: network.Standard},
-			NatGatewayPropertiesFormat: &network.NatGatewayPropertiesFormat{
-				PublicIPAddresses: &[]network.SubResource{
-					{
-						ID: to.StringPtr(azure.PublicIPID(s.Scope.SubscriptionID(), s.Scope.ResourceGroup(), natGatewaySpec.NatGatewayIP.Name)),
-					},
-				},
-			},
-		}
-		err = s.client.CreateOrUpdate(ctx, s.Scope.ResourceGroup(), natGatewaySpec.Name, natGatewayToCreate)
+		natGateway, err := async.CreateResource(ctx, s.Scope, s.client, natGatewaySpec, serviceName)
```

Comment:
```suggestion
		result, err := async.CreateResource(ctx, s.Scope, s.client, natGatewaySpec, serviceName)
```

consider renaming the error so it doesn't collide. I've used `resErr` in some other services.

Reply:
Do you mean colliding with the return value of `async.CreateResource`? I thought `natGateway` would be more specific than `result`.